### PR TITLE
Deprecate the use of pkg/errors

### PR DESCRIPTION
**Pull request by raisaat**
Deprecate the use of pkg/errors in MCO to use built-in Golang error-wrapping

Signed-off-by: Raisaat Rashid <rarashid@redhat.com>

**- What I did**

Changed all the code instances in machine-config-operator using pkg/errors to use built-in Golang error-wrapping instead.

**- How to verify it**

Make sure all the e2e and unit tests run properly.

**- Description for the changelog**

Deprecate the use of pkg/errors in machine-config-operator to use built-in Golang error-wrapping

### CODE REVIEW

**Comment by sinnykumari** [comment]
You don't need to run test again. When you push any new changes, all tests will automatically rerun

File: pkg/daemon/pivot/utils/run.go
Comment (cheesesashimi) [inline pkg/daemon/pivot/utils/run.go:53]: This is where the unit test failure is occurring. Basically, this `fmt.Errorf` requires four formatting arguments, but only one was passed in.

This should also fix the lint failure as well.

File: test/e2e-single-node/sno_mcd_test.go
Comment (cheesesashimi) [inline test/e2e-single-node/sno_mcd_test.go:410]:
```suggestion
			return fmt.Errorf("unknown error occurred %v: %w", lastErr, err)
```

**Review by cheesesashimi — CHANGES_REQUESTED** [review]
Overall, this looks great! I just have a couple of minor comments / suggestions. The biggest one is the fix for the unit tests / linter issues.

Also, I think we may be missing a few cases:
```console
$ rg -g "\!vendor" -g "*.go" --files-with-matches -F "github.com/pkg/errors"
pkg/server/api.go
pkg/daemon/on_disk_validation.go
```

Once those are addressed, we should be able to completely remove `github.com/pkg/errors`, though the actual removal should be done in a separate commit in this PR for easier review since it's vendored.

File: pkg/controller/node/node_controller.go
Comment (cheesesashimi) [inline pkg/controller/node/node_controller.go:185]: golangci-lint caught this. The reason why this was never caught before is most likely because the linter was not looking for the `goerrs.Wrap()` function. When this was changed to use `fmt.Errorf()`, the linter picked up on the pattern and alerted to an unused result since `fmt.Errorf()` is likely a more well-known signature.

That explanation aside, it's interesting that this error is never bubbled out of this function. This could be because this function is passed into a [ResourceEventHandlerFuncs](https://github.com/kubernetes/client-go/blob/master/tools/cache/controller.go#L218-L226) struct whose type does not accept a function with any return arguments. The same pattern can be seen throughout `ctrl.checkMasterNodesOnDelete`, `ctrl.reconcileMaster` and `ctrl.reconcileMasters` (all within this file, `pkg/controller/node/node_controller.go`).

As for a proposed remedy, I offer two suggestions:
1. We replace these specific calls to `fmt.Errorf()` with calls to `glog.Errorf()` so the errors are logged. This remedy would not introduce any unknown behavior and may help gain insight into how often these particular errors occur.
2. Instead of just logging the error, we handle it using the [`utilruntime.HandleError()`](https://github.com/kubernetes/apimachinery/blob/master/pkg/util/runtime/runtime.go#L83-L111) function which will try again using the `rudimentaryErrorBackoff` mechanism within that package. This function appears to be used throughout this file. However, this could introduce unknown behavior.

File: pkg/daemon/on_disk_validation.go
Comment (cheesesashimi) [inline pkg/daemon/on_disk_validation.go:158]: Just an FYI: This may cause a merge conflict depending upon whether openshift/machine-config-operator#2874 is merged before this is.

File: pkg/controller/node/node_controller.go
Comment (yuqi-zhang) [reply pkg/controller/node/node_controller.go]: Interesting. I think paths like this one are very rarely considered in the first place, since we don't generally consider node deletion operations. In fact I am not sure whether we have any testing for this case, and whether we'd be able to auto recover, if at all.

I think right now I am leaning towards option 1, since I think it's fine to increase error logging in general in case these situations do happen. For 2, I don't know how we are handling the errors at all, so I am not sure what behaviour the HandleError actually exhibits for our use case (again, they should very rarely fire)

Comment (raisaat) [reply pkg/controller/node/node_controller.go]: Should we leave option 2 for now? I have made a commit for option 1. I can go ahead and squash it with the previous commits (after a rebase since there are merge conflicts).

File: pkg/daemon/config_drift_monitor.go
Comment (cheesesashimi) [inline pkg/daemon/config_drift_monitor.go:20]: This comment is intended for other reviewers to explain why this particular change was done: We discovered that the way we were using `errors.As()` in this case was incorrect. Wrapping the error in a struct like this is the preferred way to represent a simple custom error type as opposed to doing something like `type customError error`.

**Comment by kikisdeliveryservice** [comment]
@raisaat is this the complete deprecation  in the MCO repo? The message in the PR description only mentions /cmd but this is more that that dir but at same time is this full repo?

**Comment by raisaat** [comment]
@kikisdeliveryservice sorry, I had that up when I first opened the PR. Updated it now.

**Comment by raisaat** [comment]
> @raisaat is this the complete deprecation in the MCO repo? The message in the PR description only mentions /cmd but this is more that that dir but at same time is this full repo?

@kikisdeliveryservice yes, it's the entire repo

**Review by kikisdeliveryservice — COMMENTED** [review]
a few comments, great work!

File: pkg/daemon/pivot/utils/run.go
Comment (cheesesashimi) [reply pkg/daemon/pivot/utils/run.go:51]: Ahh good catch! This should probably be a two nested errors.

File: pkg/controller/node/node_controller.go
Comment (raisaat) [reply pkg/controller/node/node_controller.go:372]: Of course! Should I make the error messages in the rest of the repo lower too?

File: pkg/daemon/pivot/utils/run.go
Comment (cheesesashimi) [reply pkg/daemon/pivot/utils/run.go:51]: Reasoning about this particular case, lets use an [aggregated error](https://github.com/kubernetes/apimachinery/blob/master/pkg/util/errors/errors.go#L46) to represent the combination of `err` and `lastErr`. It should become something like this:

```golang
import (
  kubeErrs "k8s.io/apimachinery/pkg/util/errors"
)

// ...

errs := kubeErrs.NewAggregate([]error{err, lastErr})
return "", fmt.Errorf("failed to run command %s (%d tries): %v: %w", command, backoff.Steps, errs)
```

You shouldn't have to do anything to use that package because the MCO already uses it. See: https://github.com/openshift/machine-config-operator/blob/master/test/e2e-shared-tests/helpers.go#L219-L222 for an example.

File: pkg/daemon/controlplane.go
Comment (kikisdeliveryservice) [inline pkg/daemon/controlplane.go:71]:
```suggestion
		return fmt.Errorf("Failed to set per-object-fsync for ostree: %w", err)
```
```suggestion
		return fmt.Errorf("failed to set per-object-fsync for ostree: %w", err)
```

File: pkg/controller/node/node_controller.go
Comment (kikisdeliveryservice) [reply pkg/controller/node/node_controller.go:372]: yes, if you could fix in this file, controlplane.go, on_disk_validation.go and 2 in update.go. It's not many occurances, so would be great.

File: pkg/daemon/pivot/utils/run.go
Comment (kikisdeliveryservice) [reply pkg/daemon/pivot/utils/run.go:51]: Thanks Zack!  

@raisaat I think there are a few more following this pattern in the files in this PR

Comment (raisaat) [reply pkg/daemon/pivot/utils/run.go:51]: @cheesesashimi I can aggregate the errors using `kubeErrs` in [pkg/controller/node/node_controller.go](https://github.com/openshift/machine-config-operator/pull/2868/commits/8a7c5b44e5f4567a3d2b872488db3c571f25cc34#diff-92f10e93215b8da12723d913e0985568114aab4b8afb67541c2a3a9ec5cafba6L775-R794), [pkg/daemon/drain.go](https://github.com/openshift/machine-config-operator/pull/2868/commits/8a7c5b44e5f4567a3d2b872488db3c571f25cc34#diff-58db213cf6f6259a621f8d3421b5e8af55ed71c9f52c15b255f66f0b1e78a946R69), [pkg/daemon/update.go](https://github.com/openshift/machine-config-operator/pull/2868/commits/8a7c5b44e5f4567a3d2b872488db3c571f25cc34#diff-349c0748c3f52201852d3027c29daf618b45300b949482728df6666a3f9ba245), [pkg/operator/sync.go](https://github.com/openshift/machine-config-operator/pull/2868/commits/8a7c5b44e5f4567a3d2b872488db3c571f25cc34#diff-3e205577bec1a1d1711df8bfeff30e63f044b24cfbdc69bbd7d0052fdc881891R655-R665) and [test/e2e-single-node/sno_mcd_test.go](https://github.com/openshift/machine-config-operator/pull/2868/commits/8a7c5b44e5f4567a3d2b872488db3c571f25cc34#diff-8771ad0c06d4ca41d8201443cd09bc24541cedd0176bd013ced9c0e271b958feR408-R410) too right?

File: pkg/daemon/update.go
Comment (raisaat) [inline pkg/daemon/update.go:182]: @cheesesashimi not sure if this is the right way to nest the two errors

File: test/e2e-single-node/sno_mcd_test.go
Comment (raisaat) [inline test/e2e-single-node/sno_mcd_test.go:410]: @cheesesashimi the nesting of errors here as well

File: pkg/daemon/update.go
Comment (raisaat) [inline pkg/daemon/update.go:182]: @cheesesashimi not sure if this is the right way to nest these two errors

File: test/e2e-single-node/sno_mcd_test.go
Comment (raisaat) [inline test/e2e-single-node/sno_mcd_test.go:410]: @cheesesashimi the nesting of errors here as well

File: pkg/daemon/pivot/utils/run.go
Comment (cheesesashimi) [reply pkg/daemon/pivot/utils/run.go]: Yes, you can do the same thing in those places as well. I never realized how often we used the pattern of having potentially multiple errors being returned! Anyway, by using the Aggregate error type, they'll appear to be a single error for checking purposes (e.g., `if err != nil { //... }`), but the individual errors will remain separate and can be unwrapped later, if necessary.

File: test/e2e-single-node/sno_mcd_test.go
Comment (cheesesashimi) [reply test/e2e-single-node/sno_mcd_test.go:410]: Same thing here, do something like:

```golang
errs := kubeErrs.NewAggregate([]error{lastErr, err})
if err == wait.ErrWaitTimeout {
  return fmt.Errorf("pool %s is still not updated, waited %v: %w", pool, time.Since(startTime), errs)
}
return fmt.Errorf("unknown error occurred: %w", errs)
```

File: pkg/controller/node/node_controller.go
Comment (cheesesashimi) [inline pkg/controller/node/node_controller.go:374]: Out of curiosity, why did you add these `glog.Error(err)` statements?

File: pkg/daemon/pivot/utils/run.go
Comment (cheesesashimi) [inline pkg/daemon/pivot/utils/run.go:55]: Shouldn't this be `errs := kubeErrs.NewAggregate([]error{err, lastErr})`?

File: pkg/daemon/update.go
Comment (cheesesashimi) [inline pkg/daemon/update.go:1137]: Why are you aggregating the same error twice? You should be able to use `fmt.Errorf` since there's only one error here.

File: pkg/daemon/update.go
Comment (cheesesashimi) [inline pkg/daemon/update.go:1197]: Even though there are two `err` variables here, in the case where you're aggregating both of them, one of them will shadow the other. The error that matters in this case is the one returned by `os.Remove` since the error from `os.Stat` will always be nil. I suggest something like this instead:

```golang
if _, err := os.Stat(noOrigFileStampName(f.Path)); err == nil {
  if delErr := os.Remove(noOrigFileStampName(f.Path)); delErr != nil {
    return fmt.Errorf("deleting noorig file stamp %q: %w", noOrigFileStampName(f.Path), delErr)
  }
}
```

This way, it's obvious that the delete error is what is being returned.

File: pkg/daemon/update.go
Comment (cheesesashimi) [inline pkg/daemon/update.go:1230]: Same as above because the stat error here will never not be nil.

File: pkg/daemon/update.go
Comment (cheesesashimi) [inline pkg/daemon/update.go:1270]: Same as above. We're only interested in the error from `os.Remove` here.

File: pkg/daemon/update.go
Comment (cheesesashimi) [inline pkg/daemon/update.go:1304]: Same as above. We're only interested in the error from `os.Remove`

File: pkg/daemon/update.go
Comment (cheesesashimi) [inline pkg/daemon/update.go:1595]: Same as above. However, in this case, you can aggregate the error from `os.Stat`.

File: pkg/daemon/update.go
Comment (cheesesashimi) [inline pkg/daemon/update.go:1613]: Same as above. We're only interested in the error from `os.MkdirAll` here.

File: pkg/controller/node/node_controller.go
Comment (raisaat) [reply pkg/controller/node/node_controller.go:374]: @cheesesashimi I think this relates to the discussion here: https://github.com/openshift/machine-config-operator/pull/2868#discussion_r770073515

Went with option 1.

Comment (cheesesashimi) [reply pkg/controller/node/node_controller.go:374]: Ahhh yes, that's right. Sorry!

File: pkg/daemon/pivot/utils/run.go
Comment (raisaat) [reply pkg/daemon/pivot/utils/run.go:55]: @cheesesashimi Hmm, looking back at the original code, seems like the author was not wrapping `lastErr` here. Should we still wrap the last error here or should the original line of code be replaced with something like:
```
return "", fmt.Errorf("failed to run command %s (%d tries): %w", err)
```

Comment (cheesesashimi) [reply pkg/daemon/pivot/utils/run.go:55]: I don't think it matters too much in this case. Let's go with what was there before.

File: pkg/operator/sync.go
Comment (cheesesashimi) [inline pkg/operator/sync.go:763]: Same as above. Let's aggregate `err` and `lastErr`.

File: pkg/operator/sync.go
Comment (cheesesashimi) [inline pkg/operator/sync.go:797]: Same as above, lets aggregate `err` and `lastErr`.

File: pkg/operator/sync.go
Comment (cheesesashimi) [inline pkg/operator/sync.go:848]: Same as above. Let's aggregate `err` and `lastErr`.

File: pkg/operator/sync.go
Comment (raisaat) [reply pkg/operator/sync.go:724]: Ahh thanks for catching that! Pushing a commit with the changes in.

**Comment by kikisdeliveryservice** [comment]
@raisaat lmk when this is ready

/assign

**Review by kikisdeliveryservice — APPROVED** [review]
Excellent work @raisaat !!!

Will let @cheesesashimi do a final review for his lgtm

/assin @cheesesashimi

**Comment by raisaat** [comment]
> Excellent work @raisaat !!!
> 
> Will let @cheesesashimi do a final review for his lgtm
> 
> /assign @cheesesashimi

@kikisdeliveryservice Thanks a lot, Kirsten!

**Comment by kikisdeliveryservice** [comment]
hmm so the failures seem related to this PR

see: https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_machine-config-operator/2868/pull-ci-openshift-machine-config-operator-master-e2e-agnostic-upgrade/1514370580923551744

```
 # github.com/openshift/machine-config-operator/pkg/daemon
pkg/daemon/daemon.go:703:12: undefined: "errors".Wrapf
pkg/daemon/daemon.go:706:11: undefined: "errors".Wrapf
pkg/daemon/daemon.go:713:10: undefined: "errors".Wrapf
pkg/daemon/daemon.go:721:10: undefined: "errors".Wrapf
pkg/daemon/daemon.go:726:10: undefined: "errors".Wrapf
pkg/daemon/daemon.go:733:10: undefined: "errors".Wrapf
pkg/daemon/daemon.go:738:10: undefined: "errors".Wrapf
pkg/daemon/daemon.go:744:10: undefined: "errors".Wrapf
pkg/daemon/daemon.go:752:10: undefined: "errors".Wrapf
pkg/daemon/daemon.go:758:11: undefined: "errors".Wrapf
pkg/daemon/daemon.go:758:11: too many errors 
```

/hold

**Comment by kikisdeliveryservice** [comment]
@raisaat I think you need to rebase onto master and then search for new "errors." that were just merged today via https://github.com/openshift/machine-config-operator/pull/3068

**Comment by yuqi-zhang** [comment]
Ah apologies, I duplicate some old code for that PR, and forgot to update error references.

**Comment by raisaat** [comment]
@kikisdeliveryservice Thanks! Fixed them.

@yuqi-zhang not a problem :)

**Comment by kikisdeliveryservice** [comment]
Thanks @raisaat 

/lgtm
/hold cancel

**Comment by kikisdeliveryservice** [comment]
Overriding the agnostic-upgrade job which keeps failing due to unrelated bugs and to prevent other PRs from merging which will require this to be updated again.

/override ci/prow/e2e-agnostic-upgrade

**Pull request merged**
